torch/csrc/autograd/compiled_autograd.h

```diff
@@ -0,0 +1,583 @@
+    inputs.emplace_back(tensor);
+  }
+
+  void add_set_grad_target(const at::Tensor& tensor) {
+    set_grad_targets.emplace_back(tensor);
+  }
+
+  void add_size_input(const c10::SymInt& s) {
+    all_size_inputs.emplace_back(SizeInput(default_dyn_type, s.expect_int()));
+  }
+
+  int emplace_hook(c10::SafePyObject&& fn) {
+    hooks.emplace_back(std::move(fn));
+    return hooks.size() - 1;
+  }
+
+  std::vector<SizeInput> all_size_inputs;
+  std::vector<int64_t> dyn_size_inputs;
+  std::vector<at::Tensor> inputs;
+  std::vector<at::Tensor> set_grad_targets;
+  std::vector<c10::SafePyObject> hooks;
+  bool accumulate_grad;
+  SizeInput::DynType default_dyn_type;
+};
+
+class CompiledNodeArgs {
+  // CompiledNodeArgs builds a representation of the constant values found
+  // across all the nodes in the compiled graph, via 'collect' overloads. The
+  // collected constants are specialized on by concatenation into a cache key.
+  // Tensor, symint arguments (which are lifted to become graph inputs rather
+  // than specialized on) are forwarded to the compiler and not included in the
+  // key.
+ public:
+  void collect(const at::Tensor& t) {
+    collect(t.defined());
+    if (t.defined()) {
+      _compiler.add_tensor_input(t);
+    }
+  }
+  void collect(const SavedVariable& t) {
+    collect(t.unpack(_node_call.node));
+  }
+  void collect(const c10::SymInt& t) {
+    _compiler.add_size_input(t);
+  }
+  template <typename T>
+  void collect(const std::vector<T>& t) {
+    collect_size(t.size());
+    for (size_t i = 0; i < t.size(); ++i) {
+      collect(t[i]);
+    }
+  }
+  template <typename T>
+  void collect(const c10::ArrayRef<T>& t) {
+    collect_size(t.size());
+    for (size_t i = 0; i < t.size(); ++i) {
+      collect(t[i]);
+    }
+  }
+  template <typename T>
+  void collect(const c10::OptionalArray<T>& t) {
+    collect(t.list);
+  }
+  template <typename T>
+  void collect(const c10::optional<T>& t) {
+    specialize_on_bytes(t.has_value());
+    if (t.has_value()) {
+      collect(*t);
+    }
+  }
+  template <typename A, typename B>
+  void collect(const std::pair<A, B>& t) {
+    collect(t.first);
+    collect(t.second);
+  }
+  void collect(const c10::ScalarType& t) {
+    specialize_on_bytes(t);
+  }
+  void collect(const c10::Scalar& t) {
+    auto type = t.type();
+    specialize_on_bytes(type);
+    if (type == c10::ScalarType::Double) {
+      collect(t.toDouble());
+    } else if (type == c10::ScalarType::Long) {
+      collect(t.toLong());
+    } else if (type == c10::ScalarType::Bool) {
+      collect(t.toBool());
+    } else if (type == c10::ScalarType::ComplexDouble) {
+      auto c = t.toComplexDouble();
+      collect(c.real());
+      collect(c.imag());
+    } else {
+      TORCH_CHECK(false);
+    }
+  }
+  void collect(const c10::TensorOptions& t) {
+    // TODO(jansel): are there any pointers in this type we shouldn't memcmp
+    // I think this one is wrong.... should fix it
```

Comment:
There are no pointer types; TypeMeta, the primary problematic one, is stored as an index. Depending on how `specialize_on_bytes` works, however, you may need to make sure you don't key on garbage data between fields.